controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -966,45 +966,48 @@ func (r *BareMetalHostReconciler) actionPreparing(prov provisioner.Provisioner,
 	// is invalid and needs to be reconfigured.
 	if info.host.Status.ErrorType == metal3v1alpha1.PreparationError {
 		prepareData.ActualRAIDConfig = nil
-		dirty = true
+		bmhDirty = true
 	}
 
 	// Use settings in hostFirmwareSettings if available
-	hfs, err := r.getHostFirmwareSettings(info)
+	hfsDirty, hfs, err := r.getHostFirmwareSettings(info)
 	if err != nil {
 		// wait until hostFirmwareSettings are ready
 		return actionContinue{subResourceNotReadyRetryDelay}
 	}
-	if hfs != nil {
+	if hfsDirty {
 		prepareData.ActualFirmwareSettings = hfs.Status.Settings.DeepCopy()
 		prepareData.TargetFirmwareSettings = hfs.Spec.Settings.DeepCopy()
 	}
 
-	provResult, started, err := prov.Prepare(prepareData, dirty)
+	provResult, started, err := prov.Prepare(prepareData, bmhDirty || hfsDirty)
```

Comment:
This is a change in behaviour... IIRC the flag causes us to start a new cleaning.

Previously we keep passing true until it returns started=true, at which point we save the provisioning settings and bmhDirty becomes false.
It looks to me like hfsDirty will continue to return true even after cleaning is started? I fear this will mess with the error handling.

Reply:
> This is a change in behaviour... IIRC the flag causes us to start a new cleaning.
> 
> Previously we keep passing true until it returns started=true, at which point we save the provisioning settings and bmhDirty becomes false. It looks to me like hfsDirty will continue to return true even after cleaning is started? I fear this will mess with the error handling.

Good point. I added another condition to the HFS to track whether cleaning has started and that will allow the error to be returned if Ironic fails the cleaning step. As we talked about using the condition prevents needed to add another field to the CRD.